yt_dlp/downloader/fragment.py

```diff
@@ -21,6 +23,77 @@
 )
 
 
+class Timer:
+    TIMER_FUNC = time.monotonic
+
+    def __init__(self):
+        self._last_ts = self.TIMER_FUNC()
+
+    def elapsed_seconds(self) -> float:
+        return self.TIMER_FUNC() - self._last_ts
+
+    def has_elapsed(self, seconds: float) -> bool:
+        assert seconds > 0.0
+        elapsed_seconds = self.elapsed_seconds()
+        if elapsed_seconds < seconds:
+            return False
+
+        self._last_ts += elapsed_seconds - elapsed_seconds % seconds
+        return True
+
+
+class SpeedometerMA:
+    TIMER_FUNC = time.monotonic
+    UPDATE_TIMESPAN_S = 1.0
+    AVERAGE_TIMESPAN_S = 5.0
+
+    def __init__(self, initial_bytes=0):
+        self.ts_data = [(self.TIMER_FUNC(), initial_bytes)]
+        self.timer = Timer()
+        self.last_value = None
+
+    def __call__(self, byte_counter):
+        time_now = self.TIMER_FUNC()
+
+        # only append data older than 50ms
+        if time_now - self.ts_data[-1][0] > 0.05:
+            self.ts_data.append((time_now, byte_counter))
+
+        # remove older entries
+        idx = max(0, bisect(self.ts_data, (time_now - self.AVERAGE_TIMESPAN_S, )) - 1)
+        self.ts_data[0:idx] = ()
+
+        diff_time = time_now - self.ts_data[0][0]
+        speed = (byte_counter - self.ts_data[0][1]) / diff_time if diff_time else None
+        if self.timer.has_elapsed(seconds=self.UPDATE_TIMESPAN_S):
+            self.last_value = speed
+
+        return self.last_value or speed
+
+
+class SmoothETA:
+    def __init__(self):
+        self.last_value = None
+
+    def __call__(self, value):
+        if value <= 0:
+            return 0
+
+        time_now = time.monotonic()
+        if self.last_value:
+            predicted = self.last_value - time_now
+            if predicted <= 0:
+                deviation = float('inf')
```

Comment:
Maybe use `math.inf` instead? Maybe it's just me that doesn't like `float('inf')`